x/wasm-storage/types/filters.go

```diff
@@ -0,0 +1,235 @@
+package types
+
+import (
+	"bytes"
+	"encoding/binary"
+	"reflect"
+	"slices"
+	"strconv"
+
+	"golang.org/x/exp/constraints"
+)
+
+type Filter interface {
+	// ApplyFilter takes in a list of reveals and returns an outlier
+	// list, whose value at index i indicates whether i-th reveal is
+	// an outlier. Value of 1 indicates an outlier, and value of 0
+	// indicates a non-outlier reveal.
+	ApplyFilter(reveals []RevealBody) ([]int, error)
+}
+
+type FilterNone struct{}
+
+// NewFilterNone constructs a new FilterNone object.
+func NewFilterNone(_ []byte) (FilterNone, error) {
+	return FilterNone{}, nil
+}
+
+// FilterNone declares all reveals as non-outliers.
+func (f FilterNone) ApplyFilter(reveals []RevealBody) ([]int, error) {
+	return make([]int, len(reveals)), nil
+}
+
+type FilterMode struct {
+	dataPath string // JSON path to reveal data
+}
+
+// NewFilterMode constructs a new FilerMode object given a filter
+// input.
+// Mode filter input looks as follows:
+// 0             1                  9       9+data_path_length
+// | filter_type | data_path_length |   data_path   |
+func NewFilterMode(input []byte) (FilterMode, error) {
+	var filter FilterMode
+	if len(input) < 9 {
+		return filter, ErrInvalidFilterInputLen.Wrapf("%d < %d", len(input), 9)
+	}
```

Comment:
Do we have a test for when you input a valid length but an empty string as data_path? 
According to the JSON path spec you need at least `$` in order to count as a valid data_path. But if it's handled correctly it should be fine.

Reply:
After a lil discussion with @mariocao it seems that the case with an empty data_path is erroring in the parsing of the JSON + path. However it would be better if we errored before and not execute the tally at all.

Reply:
If the data_path is invalid, tally won't be executed due to corrupt reveal error. You're right that we don't validate the data_path except its length until we parse the reveals, but I don't think we can do anything without looking at the reveal data?

Reply:
Weirdly enough, our gjson package fails when you put `$` in the beginning of a JSON path.

Reply:
The package we use for json path search seems to use a different variant of path syntax https://github.com/tidwall/gjson/blob/master/SYNTAX.md
Will check if there is a good a alternative

Reply:
Sorry ignore my comment about the $. Its indeed not a strict requirement. The gjson syntax is indeed not the same as JSONPath

Reply:
I think using JSONPath makes sense though, it's the closest we have to a standard for this kind of thing. Looks like there are some good Golang implementations

Reply:
Agree with @Thomasvdam but this change is not as trivial as I thought. GJSON was quite convenient for what we want, which is to just get string representations of JSON values at a given path. (We later parse these strings to different data types)

Reply:
Chose the package [OjG](https://github.com/ohler55/ojg) for standardized JSONPath as described [here](https://goessner.net/articles/JsonPath/).